add missing requirements for internal file pointers (#1134)
php-fig/fig-standards: accepted/PSR-17-http-factory-meta.md

stream, and therefore also assumes responsibility for the read/write pointer
state of the resource and resulting stream.

And finally, the instances created by the `RequestFactoryInterface()`,

> `ResponseFactoryInterface()` and `ServerRequestFactoryInterface` factories

Reviewer: These are not functions.

Author: It says "factories", not "functions".

There's only one function in each interface - spelling out the method-names as well seemed a bit excessive?

I'll change it if you insist though.

Reviewer: I mean `RequestFactoryInterface()` (etc) should be `RequestFactoryInterface`.

Author: Oh whoops 😂

> must generate instances that return an empty body-stream in read and write

Reviewer: I think the wording could be improved here:

> must generate instances that contain an empty `StreamInterface`, in read/write mode, as the body

Reviewer: The same goes for other places where this wording is used.

Author: Implementations of PSR-7/17 should defer the `StreamInterface` instance creation until `getBody()` is called for the first time, since you may otherwise run out of OS file handles. (Linux has a limit of 1000 by default.)

So it doesn't have to "contain" a stream instance, and probably shouldn't - it just has to "return" one.

I had initially written "read/write" as you suggest, but changed it to "read *and* write" for clarity, since the "/" could be interpreted as "or", which would be incorrect.

But I will change the wording from "body-stream" to `StreamInterface` everywhere.

- [shadowhand Dependency Inversion and PSR-7 Bodies](http://shadowhand.me/dependency-inversion-and-psr-7-bodies/)
- [PHP-FIG mailing list thread discussing factories](https://groups.google.com/d/msg/php-fig/G5pgQfQ9fpA/UWeM1gm1CwAJ)
- [PHP-FIG mailing list thread feedback on proposal](https://groups.google.com/d/msg/php-fig/piRtB2Z-AZs/8UIwY1RtDgAJ)

## 9. Errata

This recommendation initially omitted the following requirements.

### 9.1 `StreamFactoryInterface::createStream()`

> The state of the created temporary resource was unspecified - per section 5.7, the
> requirement was added to position the temporary resource at the beginning of the stream.

Reviewer: Per my other comment, this should read:

```suggestion
The state of the created temporary resource was unspecified - per section 5.7,
the recommendation was added to position the pointer consistent with an
`fwrite()` operation.
```

### 9.2 `StreamFactoryInterface::createStreamFromFile()`

> The state of the created temporary resource was unspecified - per section 5.7, the
> requirement was added to position the resource consistently with `fopen()`.

Reviewer: ```suggestion
The state of the created temporary resource was unspecified - per section 5.7,
the recommendation was added to position the pointer of the resource consistently
with `fopen()`.
```

### 9.3 `StreamFactoryInterface::createStreamFromResource()`

> The requirements did not specify whether this method might affect the state of the given
> resource - per section 5.7, the requirement was added to clarify that this method MUST NOT
> modify the position of the given resource.

Reviewer: ```suggestion
The requirements did not specify whether this method might affect the state of the given
resource - per section 5.7, the recommendation was added to not modify the pointer position
of the given resource.
```
